assets/src/edit-story/components/header/buttons.js

```diff
@@ -55,19 +59,107 @@ function PreviewButton() {
       story: { link },
     },
+    actions: { saveStory },
   } = useStory();
 
+  const [previewLinkToOpenViaDialog, setPreviewLinkToOpenViaDialog] = useState(
+    null
+  );
+
   /**
    * Open a preview of the story in current window.
    */
   const openPreviewLink = () => {
     const previewLink = addQueryArgs(link, { preview: 'true' });
-    window.open(previewLink, 'story-preview');
+
+    // Start a about:blank popup with waiting message until we complete
+    // the saving operation. That way we will not bust the popup timeout.
+    let popup;
+    try {
+      popup = window.open('about:blank', PREVIEW_TARGET);
+      if (popup) {
+        popup.document.write('<!DOCTYPE html><html><head>');
+        popup.document.write('<title>');
+        popup.document.write(
+          escapeHTML(__('Generating the preview...', 'web-stories'))
+        );
+        popup.document.write('</title>');
+        popup.document.write('</head><body>');
+        // Output "waiting" message.
+        popup.document.write(
+          escapeHTML(
+            __('Please wait. Generating the preview...', 'web-stories')
+          )
+        );
+        // Force redirect to the preview URL after 5 seconds. The saving tab
+        // might get frozen by the browser.
+        popup.document.write(
+          `<script>
+            setTimeout(function() {
+              location.replace(${JSON.stringify(previewLink)});
+            }, 5000);
+          </script>`
+        );
+      }
+    } catch (e) {
+      // Ignore errors. Anything can happen with a popup. The errors
+      // will be resolved after the story is saved.
+    }
+
+    // @todo: See https://github.com/google/web-stories-wp/issues/1149. This
+    // has the effect of pushing the changes to a published story.
+    saveStory().then(() => {
+      let previewOpened = false;
+      if (popup && !popup.closed) {
+        try {
+          // The `popup.location.href` will fail if the expected window has
+          // been naviagted to a different origin.
+          if (popup.location.href) {
+            popup.location.replace(previewLink);
+            previewOpened = true;
+          }
+        } catch (e) {
+          // Ignore the errors. They will simply trigger the "try again"
+          // dialog.
+        }
+      }
+      if (!previewOpened) {
+        setPreviewLinkToOpenViaDialog(previewLink);
+      }
+    });
   };
+
+  const openPreviewLinkSync = (evt) => {
+    setPreviewLinkToOpenViaDialog(null);
+    // Ensure that this method is as safe as possible and pass the random
+    // target in case the normal target is not openable.
+    window.open(previewLinkToOpenViaDialog, PREVIEW_TARGET + Math.random());
+    evt.preventDefault();
+  };
+
   return (
-    <Outline onClick={openPreviewLink} isDisabled={isSaving}>
-      {__('Preview', 'web-stories')}
-    </Outline>
+    <>
+      <Outline onClick={openPreviewLink} isDisabled={isSaving}>
+        {__('Preview', 'web-stories')}
```

Comment:
I wonder if we should change `Preview` to `View` or `Save & Preview` or `Save & View` or something similar to reflect more accurately what the button does until the actual Preview is done. Thoughts?

Reply:
Yeah, it's not really "preview", if it's the same as everyone else is seeing.